File: src/lib/api/groups.ts
import { invoke } from "@tauri-apps/api/core";

import type { RowPage } from "./rows";
import type { RowSelection } from "./types";

export interface GroupSummary {
  id: number;
  name: string;
  memberCount: number;
  createdAt: string;
}

export function createGroup(name: string): Promise<GroupSummary> {
  return invoke<GroupSummary>("create_group", { name });
}

export function renameGroup(groupId: number, newName: string): Promise<GroupSummary> {
  return invoke<GroupSummary>("rename_group", { groupId, newName });
}

export function deleteGroup(groupId: number): Promise<boolean> {
  return invoke<boolean>("delete_group", { groupId });
}

export function deleteEmptyGroups(): Promise<number> {
  return invoke<number>("delete_empty_groups");
}

export function listGroups(): Promise<GroupSummary[]> {
  return invoke<GroupSummary[]>("list_groups");
}

export function assignRowsToGroup(
  selection: RowSelection,
  groupId: number,
): Promise<number> {
  return invoke<number>("assign_rows_to_group", { selection, groupId });
}

export function ungroupRows(selection: RowSelection): Promise<number> {
  return invoke<number>("ungroup_rows", { selection });
}

export type SimilarityMode = "artists" | "positivePrompt";

export interface SuggestedGroup {
  name: string;
  rowIds: number[];
}

export function suggestGroups(
  mode: SimilarityMode,
  threshold: number,
): Promise<SuggestedGroup[]> {
  return invoke<SuggestedGroup[]>("suggest_groups", { mode, threshold });
}

export function getGroupMembers(
  groupId: number,
  offset: number,
  limit: number,
): Promise<RowPage> {
  return invoke<RowPage>("get_group_members", { groupId, offset, limit });
}

File: src-tauri/src/app/runtime/groups.rs
use super::{AppRuntime, AppRuntimeError};
use crate::db::{GroupSummary, RowPage, RowSelection};

impl AppRuntime {
    pub(crate) fn create_group(&self, name: &str) -> Result<GroupSummary, AppRuntimeError> {
        self.with_database_mut(|db| db.create_group(name))
    }

    pub(crate) fn restore_group(
        &self,
        group: &GroupSummary,
    ) -> Result<GroupSummary, AppRuntimeError> {
        self.with_database_mut(|db| db.restore_group(group))
    }

    pub(crate) fn rename_group(
        &self,
        group_id: i64,
        new_name: &str,
    ) -> Result<GroupSummary, AppRuntimeError> {
        self.with_database_mut(|db| db.rename_group(group_id, new_name))
    }

    pub(crate) fn delete_group(&self, group_id: i64) -> Result<bool, AppRuntimeError> {
        self.with_database_mut(|db| db.delete_group(group_id))
    }

    pub(crate) fn delete_empty_groups(&self) -> Result<u64, AppRuntimeError> {
        self.with_database_mut(|db| db.delete_empty_groups())
    }

    pub(crate) fn list_groups(&self) -> Result<Vec<GroupSummary>, AppRuntimeError> {
        self.with_database(|db| db.list_groups())
    }

    pub(crate) fn assign_rows_to_group(
        &self,
        selection: &RowSelection,
        group_id: i64,
    ) -> Result<u64, AppRuntimeError> {
        self.with_database_mut(|db| db.assign_rows_to_group(selection, group_id))
    }

    pub(crate) fn ungroup_rows(&self, selection: &RowSelection) -> Result<u64, AppRuntimeError> {
        self.with_database_mut(|db| db.ungroup_rows(selection))
    }

    pub(crate) fn get_group_members(
        &self,
        group_id: i64,
        offset: u64,
        limit: u32,
    ) -> Result<RowPage, AppRuntimeError> {
        self.with_database(|db| db.get_group_members(group_id, offset, limit))
    }
}
